File: exercises/domain-5/block-5.2/starter.ts
type ErrorClass = "retry" | "escalate" | "degrade"

interface EscalationRecord {
  step: string
  error: string
  timestamp: string
  resolution: string
}

class EscalationManager {
  escalationLog: EscalationRecord[] = []
  private failureCounts: Map<string, number> = new Map()
  private readonly RETRY_LIMIT = 3

  // TODO: Classify error as retry/escalate/degrade
  classifyError(error: Error): ErrorClass {
    // Retry: network errors, rate limits, timeouts
    // Escalate: permission errors, safety violations, data corruption
    // Degrade: optional feature failures
    return "retry"
  }

  // TODO: Run fn with error classification and escalation logic
  async handleWithEscalation<T>(
    stepName: string,
    fn: () => Promise<T>,
    options: { required: boolean } = { required: true }
  ): Promise<T | null> {
    return null
  }
}

// Demo
async function main() {
  const manager = new EscalationManager()

  // Step 1: Will succeed after 2 retries (simulated)
  let attempts = 0
  const result1 = await manager.handleWithEscalation("fetch-data", async () => {
    attempts++
    if (attempts < 3) throw new Error("ECONNREFUSED: network error")
    return { data: "success" }
  })
  console.log("Step 1 result:", result1)

  // Step 2: Permission error → immediate escalation
  const result2 = await manager.handleWithEscalation("write-db", async () => {
    throw new Error("EPERM: permission denied")
  })
  console.log("Step 2 result:", result2)

  // Step 3: Optional step failure → degrade
  const result3 = await manager.handleWithEscalation("generate-chart", async () => {
    throw new Error("Chart service unavailable")
  }, { required: false })
  console.log("Step 3 result (optional):", result3)

  console.log("\nEscalation log:", manager.escalationLog)
}

main().catch(console.error)

File: exercises/domain-5/block-5.2/starter.py
import asyncio
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import TypeVar, Callable, Awaitable

T = TypeVar("T")


@dataclass
class EscalationRecord:
    step: str
    error: str
    timestamp: str
    resolution: str


class EscalationManager:
    def __init__(self):
        self.escalation_log: list[EscalationRecord] = []
        self._failure_counts: dict[str, int] = {}
        self.RETRY_LIMIT = 3

    def classify_error(self, error: Exception) -> str:
        """TODO: Return 'retry', 'escalate', or 'degrade'"""
        return "retry"

    async def handle_with_escalation(
        self,
        step_name: str,
        fn: Callable[[], Awaitable[T]],
        required: bool = True,
    ) -> T | None:
        """TODO: Run fn with error classification and escalation"""
        return None


async def main():
    manager = EscalationManager()
    attempts = 0

    # Step 1: Succeeds after 2 retries
    async def fetch_data():
        nonlocal attempts
        attempts += 1
        if attempts < 3:
            raise ConnectionError("ECONNREFUSED: network error")
        return {"data": "success"}

    result1 = await manager.handle_with_escalation("fetch-data", fetch_data)
    print("Step 1:", result1)

    # Step 2: Permission error → immediate escalation
    async def write_db():
        raise PermissionError("EPERM: permission denied")

    result2 = await manager.handle_with_escalation("write-db", write_db)
    print("Step 2:", result2)

    # Step 3: Optional failure → degrade
    async def generate_chart():
        raise RuntimeError("Chart service unavailable")

    result3 = await manager.handle_with_escalation("generate-chart", generate_chart, required=False)
    print("Step 3 (optional):", result3)

    print("\nEscalation log:", manager.escalation_log)


asyncio.run(main())
